internal/cephfs/groupcontrollerserver.go

```diff
@@ -61,3 +72,563 @@ func (cs *ControllerServer) validateVolumeGroupSnapshotRequest(
+		// journal
+		delReq := &csi.DeleteSnapshotRequest{
+			SnapshotId: resp.GetSnapshot().GetSnapshotId(),
+			Secrets:    req.Secrets,
+		}
+		_, dErr := cs.DeleteSnapshot(ctx, delReq)
+		if dErr != nil {
+			log.ErrorLog(ctx, "failed to delete snapshot %s: %v", resp.GetSnapshot().GetSnapshotId(), dErr)
+		}
+
+		return nil, err
+	}
+
+	return resp, nil
+}
+
+// checkIfFSNeedQuiesceRelease checks that do we have snapshots for all the
+// volumes stored in the omap so that we can release the quiesce.
+func checkIfFSNeedQuiesceRelease(vgs *store.VolumeGroupSnapshotIdentifier, volIDs []string) bool {
+	if vgs == nil {
+		return false
+	}
+	// If the number of volumes in the snapshot is not equal to the number of volumes
+	snapshottedVolIDs := vgs.GetVolumeIDs()
+
+	return len(snapshottedVolIDs) == len(volIDs)
+}
+
+func getClusterIDForVolumeID(fms map[string]core.SubVolumeQuiesceClient, volumeID string) string {
+	for _, fm := range fms {
+		for _, vol := range fm.GetVolumes() {
+			if vol.VolumeID == volumeID {
+				return vol.ClusterID
+			}
+		}
+	}
+
+	return ""
+}
+
+func getFsNamesAndSubVolumeFromVolumeIDs(ctx context.Context,
+	secret map[string]string,
+	volIDs []string,
+	cr *util.Credentials) (
+	map[string]core.SubVolumeQuiesceClient,
+	error,
+) {
+	type fs struct {
+		fsName                string
+		volumes               []core.Volume
+		subVolumeGroupMapping map[string][]string
+		monitors              string
+	}
+	fm := map[string]fs{}
+	for _, volID := range volIDs {
+		// Find the volume using the provided VolumeID
+		volOptions, _, err := store.NewVolumeOptionsFromVolID(ctx,
+			volID, nil, secret, "", false)
+		if err != nil {
+			return nil, err
+		}
+		volOptions.Destroy()
+		// choosing monitorIP's and fsName as the unique key
+		// TODO: Need to use something else as the unique key as users can
+		// still choose the different monitorIP's and fsName for subvolumes
+		uniqueName := volOptions.Monitors + volOptions.FsName
+		fm[uniqueName] = fs{
+			fsName: volOptions.FsName,
+			subVolumeGroupMapping: map[string][]string{
+				volOptions.SubVolume.SubvolumeGroup: append(
+					fm[volOptions.FsName].subVolumeGroupMapping[volOptions.SubVolume.SubvolumeGroup],
+					volOptions.SubVolume.VolID),
+			},
+			volumes: append(fm[volOptions.FsName].volumes, core.Volume{
+				VolumeID:  volID,
+				ClusterID: volOptions.ClusterID,
+			}),
+			monitors: volOptions.Monitors,
+		}
+	}
+	fsk := map[string]core.SubVolumeQuiesceClient{}
+	var err error
+	for k, v := range fm {
+		conn := &util.ClusterConnection{}
+		if err = conn.Connect(v.monitors, cr); err != nil {
+			return nil, err
+		}
+		fsk[k], err = core.NewSubVolumeQuiesce(v.fsName, v.volumes, v.subVolumeGroupMapping, conn)
+		if err != nil {
+			log.ErrorLog(ctx, "failed to get subvolume quiesce: %v", err)
+
+			return nil, err
+		}
+	}
+
+	return fsk, nil
+}
+
+func matchesSourceVolumeIDs(sourceVolumeIDs, volumeIDsInOMap []string) bool {
```

Comment:
interesting that there is no standard function for this, consider using `Equal()` from `k8s.io/utils/strings/slices`?